Rules/UseUsingScopeModifierInNewRunspaces.cs

```diff
@@ -0,0 +1,375 @@
+        {
+            return string.Format(CultureInfo.CurrentCulture, Strings.SourceName);
+        }
+
+        /// <summary>
+        /// Checks if a variable is initialized and referenced in either its assignment or children scopes
+        /// </summary>
+        /// <param name="scriptBlockAst">Ast of type ScriptBlock</param>
+        /// <param name="fileName">Name of file containing the ast</param>
+        /// <returns>An enumerable containing diagnostic records</returns>
+        private IEnumerable<DiagnosticRecord> AnalyzeScriptBlockAst(ScriptBlockAst scriptBlockAst, string fileName)
+        {
+            var astsToProcess = new List<Ast>();
+            var nonAssignedNonUsingVars = new List<VariableExpressionAst>();
+
+            astsToProcess.AddRange(
+                FindAllInlineScriptAsts(scriptBlockAst));
+
+            astsToProcess.AddRange(
+                FindAllStartJobAsts(scriptBlockAst));
+
+            astsToProcess.AddRange(
+                FindAllForeachParallelAsts(scriptBlockAst));
+
+            //   If -ComputerName or -Session is not specified, you cannot use a variable with a using: scope modifier. 
+            //   Also tricky; one can open a persistent session, and use subsequent `invoke-command` calls to that, where variables
+            //   assigned in previous calls are still valid:
+            //   https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_remote_variables?view=powershell-7#using-remote-variables
+            //   Because -ComputerName and -Session parameters are in different parameter sets, we can treat them separately
+            astsToProcess.AddRange(
+                FindAllInvokeCommandComputerAsts(scriptBlockAst));
+
+            // process Invoke-Command -Session separately
+            nonAssignedNonUsingVars.AddRange(
+                ProcessInvokeCommandSessionAsts(scriptBlockAst));
+
+
+            foreach (var ast in astsToProcess)
+            {
+                if (ast == null)
+                    continue;
+
+                var varsInAssignments = FindVarsInAssignmentAsts(ast);
+                
+                nonAssignedNonUsingVars.AddRange(
+                    FindNonAssignedNonUsingVarAsts(ast, varsInAssignments));
+            }
+            
+            foreach (var variableExpression in nonAssignedNonUsingVars)
+            {
+                if (variableExpression == null)
+                    continue;
+
+                yield return new DiagnosticRecord(
+                    message: string.Format(CultureInfo.CurrentCulture,
+                        Strings.UseUsingScopeModifierInNewRunspacesError, variableExpression.ToString()),
+                    extent: variableExpression.Extent,
+                    ruleName: GetName(),
+                    severity: DiagnosticSeverity.Warning,
+                    scriptPath: fileName,
+                    ruleId: variableExpression.ToString(),

```

Comment:
Is this the correct rule ID? I thought rule ID was supposed to be the name of the rule generating the diagnostic

Reply:
No, you are right, it is not. I have changed it to the rule name.